Core/HLE/SocketManager.h

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include "Common/Net/SocketCompat.h"
+
+// Keep track of who's using a socket.
+enum class SocketState {
+	Unused,
+	UsedNetInet,
+	UsedProAdhoc,
+};
+
+// Internal socket state tracking
+struct InetSocket {
```

Comment:
We should store non-blocking state here too, and probably some other value that can be set using `SetSockOpt` too just like non-blocking state.

Reply:
Also, since we have stored the socket protocol here, we can replace these code at sceNetInet.cpp by forwarding only TCP/UDP instead of forwarding both protocol:

```c++
// Enable Port-forwarding
	// TODO: Check the socket type/protocol for SOCK_STREAM/SOCK_DGRAM or IPPROTO_TCP/IPPROTO_UDP instead of forwarding both protocol
	// InetSocket* sock = pspSockets.Get<InetSocket>(socket, error);
	// UPnP_Add((sock->type == SOCK_STREAM)? IP_PROTOCOL_TCP: IP_PROTOCOL_UDP, port, port);	
	unsigned short port = ntohs(saddr.in.sin_port);
	UPnP_Add(IP_PROTOCOL_UDP, port, port);
	UPnP_Add(IP_PROTOCOL_TCP, port, port);
```

I opened both TCP and UDP because i use the native socket directly and there is noway to retrieve the socket protocol (at least the easy way as i remembered).